### PR TITLE
refactor SFTP backend to use upstream dep, not our fork

### LINKED ISSUE #21518: SFTP transfer produces too many audit events
While transfering files over SFTP many audit events is produced for read/write action, example:

```
2023-02-08T21:04:07-05:00 INFO [AUDIT]     sftp action:3 addr.local:[::1]:3022 addr.remote:[::1]:64584 cluster_name:example.com code:TS003I ei:0 event:sftp login:jnyckowski namespace:default path:/Users/jnyckowski/teleport/teleport-v12.0.0-dev-linux-amd64-bin.tar.gz server_hostname:example.com server_id:854e9299-c604-4af8-baa9-2580c4337a84 sid: time:2023-02-09T02:04:06.217591Z uid:21bdb751-9936-4b17-b37d-fd23a4c09a23 user:bob working_directory:/Users/jnyckowski events/emitter.go:265
2023-02-08T21:04:07-05:00 INFO [AUDIT]     sftp action:3 addr.local:[::1]:3022 addr.remote:[::1]:64584 cluster_name:example.com code:TS003I ei:0 event:sftp login:jnyckowski namespace:default path:/Users/jnyckowski/teleport/teleport-v12.0.0-dev-linux-amd64-bin.tar.gz server_hostname:example.com server_id:854e9299-c604-4af8-baa9-2580c4337a84 sid: time:2023-02-09T02:04:06.217591Z uid:21bdb751-9936-4b17-b37d-fd23a4c09a23 user:bob working_directory:/Users/jnyckowski events/emitter.go:265
{"ei":0,"event":"sftp","uid":"21bdb751-9936-4b17-b37d-fd23a4c09a23","code":"TS003I","time":"2023-02-09T02:04:06.217591Z","cluster_name":"example.com","user":"bob","login":"jnyckowski","addr.local":"[::1]:3022","addr.remote":"[::1]:64584","sid":"","namespace":"default","server_id":"854e9299-c604-4af8-baa9-2580c4337a84","server_hostname":"example.com","working_directory":"/Users/jnyckowski","path":"/Users/jnyckowski/teleport/teleport-v12.0.0-dev-linux-amd64-bin.tar.gz","action":3}
2023-02-08T21:04:07-05:00 INFO [AUDIT]     sftp action:3 addr.local:[::1]:3022 addr.remote:[::1]:64584 cluster_name:example.com code:TS003I ei:0 event:sftp login:jnyckowski namespace:default path:/Users/jnyckowski/teleport/teleport-v12.0.0-dev-linux-amd64-bin.tar.gz server_hostname:example.com server_id:854e9299-c604-4af8-baa9-2580c4337a84 sid: time:2023-02-09T02:04:06.217912Z uid:18321906-e9e4-4cd9-a4b8-1ee077b64233 user:bob working_directory:/Users/jnyckowski events/emitter.go:265
2023-02-08T21:04:07-05:00 INFO [AUDIT]     sftp action:3 addr.local:[::1]:3022 addr.remote:[::1]:64584 cluster_name:example.com code:TS003I ei:0 event:sftp login:jnyckowski namespace:default path:/Users/jnyckowski/teleport/teleport-v12.0.0-dev-linux-amd64-bin.tar.gz server_hostname:example.com server_id:854e9299-c604-4af8-baa9-2580c4337a84 sid: time:2023-02-09T02:04:06.217912Z uid:18321906-e9e4-4cd9-a4b8-1ee077b64233 user:bob working_directory:/Users/jnyckowski events/emitter.go:265
{"ei":0,"event":"sftp","uid":"18321906-e9e4-4cd9-a4b8-1ee077b64233","code":"TS003I","time":"2023-02-09T02:04:06.217912Z","cluster_name":"example.com","user":"bob","login":"jnyckowski","addr.local":"[::1]:3022","addr.remote":"[::1]:64584","sid":"","namespace":"default","server_id":"854e9299-c604-4af8-baa9-2580c4337a84","server_hostname":"example.com","working_directory":"/Users/jnyckowski","path":"/Users/jnyckowski/teleport/teleport-v12.0.0-dev-linux-amd64-bin.tar.gz","action":3}
2023-02-08T21:04:07-05:00 INFO [AUDIT]     sftp action:3 addr.local:[::1]:3022 addr.remote:[::1]:64584 cluster_name:example.com code:TS003I ei:0 event:sftp login:jnyckowski namespace:default path:/Users/jnyckowski/teleport/teleport-v12.0.0-dev-linux-amd64-bin.tar.gz server_hostname:example.com server_id:854e9299-c604-4af8-baa9-2580c4337a84 sid: time:2023-02-09T02:04:06.218102Z uid:9136fffb-ac55-4dd3-a274-62f4b8cb6cb5 user:bob working_directory:/Users/jnyckowski events/emitter.go:265
2023-02-08T21:04:07-05:00 INFO [AUDIT]     sftp action:3 addr.local:[::1]:3022 addr.remote:[::1]:64584 cluster_name:example.com code:TS003I ei:0 event:sftp login:jnyckowski namespace:default path:/Users/jnyckowski/teleport/teleport-v12.0.0-dev-linux-amd64-bin.tar.gz server_hostname:example.com server_id:854e9299-c604-4af8-baa9-2580c4337a84 sid: time:2023-02-09T02:04:06.218102Z uid:9136fffb-ac55-4dd3-a274-62f4b8cb6cb5 user:bob working_directory:/Users/jnyckowski events/emitter.go:265
{"ei":0,"event":"sftp","uid":"9136fffb-ac55-4dd3-a274-62f4b8cb6cb5","code":"TS003I","time":"2023-02-09T02:04:06.218102Z","cluster_name":"example.com","user":"bob","login":"jnyckowski","addr.local":"[::1]:3022","addr.remote":"[::1]:64584","sid":"","namespace":"default","server_id":"854e9299-c604-4af8-baa9-2580c4337a84","server_hostname":"example.com","working_directory":"/Users/jnyckowski","path":"/Users/jnyckowski/teleport/teleport-v12.0.0-dev-linux-amd64-bin.tar.gz","action":3}
2023-02-08T21:04:07-05:00 INFO [AUDIT]     sftp action:3 addr.local:[::1]:3022 addr.remote:[::1]:64584 cluster_name:example.com code:TS003I ei:0 event:sftp login:jnyckowski namespace:default path:/Users/jnyckowski/teleport/teleport-v12.0.0-dev-linux-amd64-bin.tar.gz server_hostname:example.com server_id:854e9299-c604-4af8-baa9-2580c4337a84 sid: time:2023-02-09T02:04:06.219097Z uid:9207ec4c-c553-439c-8571-6efda3f03ffd user:bob working_directory:/Users/jnyckowski events/emitter.go:265
2023-02-08T21:04:07-05:00 INFO [AUDIT]     sftp action:3 addr.local:[::1]:3022 addr.remote:[::1]:64584 cluster_name:example.com code:TS003I ei:0 event:sftp login:jnyckowski namespace:default path:/Users/jnyckowski/teleport/teleport-v12.0.0-dev-linux-amd64-bin.tar.gz server_hostname:example.com server_id:854e9299-c604-4af8-baa9-2580c4337a84 sid: time:2023-02-09T02:04:06.219097Z uid:9207ec4c-c553-439c-8571-6efda3f03ffd user:bob working_directory:/Users/jnyckowski events/emitter.go:265
{"ei":0,"event":"sftp","uid":"9207ec4c-c553-439c-8571-6efda3f03ffd","code":"TS003I","time":"2023-02-09T02:04:06.219097Z","cluster_name":"example.com","user":"bob","login":"jnyckowski","addr.local":"[::1]:3022","addr.remote":"[::1]:64584","sid":"","namespace":"default","server_id":"854e9299-c604-4af8-baa9-2580c4337a84","server_hostname":"example.com","working_directory":"/Users/jnyckowski","path":"/Users/jnyckowski/teleport/teleport-v12.0.0-dev-linux-amd64-bin.tar.gz","action":3}
2023-02-08T21:04:07-05:00 INFO [AUDIT]     sftp action:3 addr.local:[::1]:3022 addr.remote:[::1]:64584 cluster_name:example.com code:TS003I ei:0 event:sftp login:jnyckowski namespace:default path:/Users/jnyckowski/teleport/teleport-v12.0.0-dev-linux-amd64-bin.tar.gz server_hostname:example.com server_id:854e9299-c604-4af8-baa9-2580c4337a84 sid: time:2023-02-09T02:04:06.219216Z uid:2fbc0a71-021d-4e62-8fdf-2a3906e200ff user:bob working_directory:/Users/jnyckowski events/emitter.go:265
2023-02-08T21:04:07-05:00 INFO [AUDIT]     sftp action:3 addr.local:[::1]:3022 addr.remote:[::1]:64584 cluster_name:example.com code:TS003I ei:0 event:sftp login:jnyckowski namespace:default path:/Users/jnyckowski/teleport/teleport-v12.0.0-dev-linux-amd64-bin.tar.gz server_hostname:example.com server_id:854e9299-c604-4af8-baa9-2580c4337a84 sid: time:2023-02-09T02:04:06.219216Z uid:2fbc0a71-021d-4e62-8fdf-2a3906e200ff user:bob working_directory:/Users/jnyckowski events/emitter.go:265
{"ei":0,"event":"sftp","uid":"2fbc0a71-021d-4e62-8fdf-2a3906e200ff","code":"TS003I","time":"2023-02-09T02:04:06.219216Z","cluster_name":"example.com","user":"bob","login":"jnyckowski","addr.local":"[::1]:3022","addr.remote":"[::1]:64584","sid":"","namespace":"default","server_id":"854e9299-c604-4af8-baa9-2580c4337a84","server_hostname":"example.com","working_directory":"/Users/jnyckowski","path":"/Users/jnyckowski/teleport/teleport-v12.0.0-dev-linux-amd64-bin.tar.gz","action":3}
```

Bug details:
- Teleport version 10.3+
- Recreation steps: Copy a file over SFTP
- Debug logs

### LINKED ISSUE #22932: Switch our SFTP to upstream
Our current implementation uses https://github.com/gravitational/sftp, which is a fork of https://github.com/pkg/sftp. Switching the SFTP client to `ServerRequest` https://github.com/pkg/sftp/blob/master/request-server.go#L24-L35 should allow us to keep the current functionality and get rid of the fork.

### PR DESCRIPTION
**Pull request by capnspacehook**
This change also greatly reduces the number of SFTP audit logs. Now SFTP events are only sent when files are opened or modified in any way, instead of for *every* SFTP request.

Fixes https://github.com/gravitational/teleport/issues/21518.
Closes https://github.com/gravitational/teleport/issues/22932.

### CODE REVIEW

**Review by strideynet — APPROVED** [review]

File: tool/teleport/common/sftp.go [hunk 1]
```diff
@@ -52,275 +75,506 @@ func (c compositeCh) Close() error {
...
+	return f, nil
+}
 
-	sftpEvents := make(chan *apievents.SFTP, 1)
-	sftpSrv, err := sftp.NewServer(ch, sftp.WithRequestCallback(func(reqPacket sftp.RequestPacket) {
-		event, ok := handleSFTPEvent(reqPacket)
-		if !ok {
-			// We don't care about this type of SFTP request, move on
+// Filecmd handles file modification requests.
+func (s *sftpHandler) Filecmd(req *sftp.Request) (retErr error) {
```
Comment (r0mant) [inline tool/teleport/common/sftp.go:166]: It's a shame we basically have to reimplement these command handlers ourselves. Are there no "default handlers" somewhere in the sftp package we could wrap to add auditing?

File: lib/sshutils/sftp/sftp.go [hunk 2]
```diff
@@ -283,9 +283,12 @@ func (c *Config) transfer(ctx context.Context) error {
 		// if there are multiple source paths and the destination path
 		// doesn't exist, create it as a directory
 		if len(c.srcPaths) > 1 {
-			if err := c.dstFS.Mkdir(ctx, c.dstPath, teleport.SharedDirMode); err != nil {
+			if err := c.dstFS.Mkdir(ctx, c.dstPath); err != nil {
 				return trace.Errorf("error creating %s directory %q: %w", c.dstFS.Type(), c.dstPath, err)
 			}
+			if err := c.dstFS.Chmod(ctx, c.dstPath, 0o755); err != nil {
```
Comment (r0mant) [inline lib/sshutils/sftp/sftp.go:289]: Use constant here.

File: lib/sshutils/sftp/local.go [hunk 3]
```diff
@@ -81,25 +81,25 @@ func (l localFS) Open(ctx context.Context, path string) (fs.File, error) {
...
 }
 
-func (l localFS) Mkdir(ctx context.Context, path string, mode os.FileMode) error {
+func (l localFS) Mkdir(ctx context.Context, path string) error {
 	if err := ctx.Err(); err != nil {
 		return err
 	}
 
-	err := os.MkdirAll(path, mode)
+	err := os.MkdirAll(path, 0o755)
```
[the remark this answers is not in the document]
Comment (capnspacehook) [reply lib/sshutils/sftp/local.go:102]: Modifications in our `sftp` fork allowed us to do that before, now we have to call `Chmod` directly after creating a file

File: lib/sshutils/sftp/sftp.go (hunk 2 above)
Comment (capnspacehook) [reply lib/sshutils/sftp/sftp.go]: Done

File: tool/teleport/common/sftp.go [hunk 4]
```diff
@@ -30,11 +32,32 @@ import (
...
 
+	"github.com/gravitational/teleport"
 	apievents "github.com/gravitational/teleport/api/types/events"
 	"github.com/gravitational/teleport/lib/events"
 	"github.com/gravitational/teleport/lib/utils"
 )
 
+const (
+	defaultFilePerms = 0o644
+	defaultDirPerms  = 0o755
```
[the remark this answers is not in the document]
Comment (capnspacehook) [reply tool/teleport/common/sftp.go]: Done

**Comment by capnspacehook** [comment]
Friendly ping @greedy52

**Review by r0mant — APPROVED** [review]

File: lib/sshutils/sftp/local.go (hunk 3 above)
[the remark this answers is not in the document]
Comment (jakule) [reply lib/sshutils/sftp/local.go]: @capnspacehook Can you backport your mkdir implementation upstream? I think that it should be accepted, and then the chmod trick won't be needed.

CC @r0mant

**Pull request merged**